kalite/contentload/management/commands/generate_assessment_zips.py

```diff
@@ -0,0 +1,116 @@
+import copy
+import json
+import os
+import re
+import requests
+import tempfile
+import zipfile
+from multiprocessing.dummy import Pool as ThreadPool
+
+from django.conf import settings
+from django.core.management.base import NoArgsCommand
+
+logging = settings.LOG
+
+ZIP_FILE_PATH = os.path.join(settings.PROJECT_PATH, "assessment_item_resources.zip")
+ASSESSMENT_ITEMS_PATH = os.path.join(settings.PROJECT_PATH, "..", "data", "khan", "assessmentitems.json")
+
+
+class Command(NoArgsCommand):
+
+    def handle_noargs(self, **options):
+        logging.info("fetching assessment items")
+
+        # load the assessmentitems
+        assessment_items = json.load(open(ASSESSMENT_ITEMS_PATH))
+
+        image_urls = all_image_urls(assessment_items)
+
+        logging.info("rewriting image urls")
+        new_assessment_items = localhosted_image_urls(assessment_items)
+
+        # TODO(jamalex): We should migrate this away from direct-to-zip so that we can re-run it
+        # without redownloading all files. Not possible currently because ZipFile has no `delete`.
+        logging.info("downloading images")
+        with open(ZIP_FILE_PATH, "w") as f:
+            zf = zipfile.ZipFile(ZIP_FILE_PATH, "w")  # zipfile.ZipFile isn't a context manager yet for python 2.6
+            write_assessment_to_zip(zf, new_assessment_items)
+            zip_file_path = download_urls(zf, image_urls)
+            zf.close()
+
+        logging.info("Zip File with images placed in %s" % zip_file_path)
+
+
+def write_assessment_to_zip(zf, assessment_items):
+    assessment_json_string = json.dumps(assessment_items)
+    zf.writestr("assessmentitems.json", assessment_json_string)
+
+
+def download_urls(zf, urls):
+
+    urls = set(urls)
+
+    pool = ThreadPool(5)
+    download_to_zip_func = lambda url: download_url_to_zip(zf, url)
+    pool.map(download_to_zip_func, urls)
+
+    return ZIP_FILE_PATH
+
+
+def download_url_to_zip(zipfile, url):
+    url = url.replace("https://ka-", "http://ka-")
```

Comment:
did someone say other sources (s)(:))....when you need an uninformed voice
to test let me know.....

On Tue, Feb 3, 2015 at 3:10 PM, benjaoming <notifications@github.com> wrote:

> In kalite/contentload/management/commands/generate_assessment_zips.py
> <https://github.com/learningequality/ka-lite/pull/2896#discussion_r24040076>
> :
>
> > +    zf.writestr("assessmentitems.json", assessment_json_string)
> > +
> > +
> > +def download_urls(zf, urls):
> > +
> > +    urls = set(urls)
> > +
> > +    pool = ThreadPool(5)
> > +    download_to_zip_func = lambda url: download_url_to_zip(zf, url)
> > +    pool.map(download_to_zip_func, urls)
> > +
> > +    return ZIP_FILE_PATH
> > +
> > +
> > +def download_url_to_zip(zipfile, url):
> > +    url = url.replace("https://ka-", "http://ka-")
>
> @jamalex <https://github.com/jamalex> -- But then maybe the code should
> say *Because KA's servers are errornously putting "https" and not "http"
> in the URLs, we make this necessary replacement in the URL before
> downloading the zip*... I mean, what's going on is not too obvious, nor
> does it seem warranted in a function that's responsible for downloading,
> not sanitizing data :) Makes more sense once we start downloading stuff
> from other sources, I guess.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/learningequality/ka-lite/pull/2896/files#r24040076>.
>